tests/DataCache.test.ts

```diff
@@ -0,0 +1,159 @@
+  });
+
+  it("calls dispose() on entries when they are evicted", () => {
+    const cache = new DataCache<DisposableString>(1);
+    const a = new DisposableString("A");
+    const b = new DisposableString("B");
+
+    cache.insert("1", a);
+    expect(a.hasBeenDisposed).toBe(false);
+
+    // Key 1 is evicted, so `a` should be disposed.
+    cache.insert("2", b);
+    expect(a.hasBeenDisposed).toBe(true);
+    expect(b.hasBeenDisposed).toBe(false);
+  });
+
+  it("moves values to front of list when they are accessed (evicts LRU)", () => {
+    const cache = new DataCache<DisposableString>(3);
+    cache.insert("1", new DisposableString("A"));
+    cache.insert("2", new DisposableString("B"));
+    cache.insert("3", new DisposableString("C"));
+
+    // 1 is oldest, but accessing it will make it the newest.
+    // 2 is now the oldest and will be evicted next time a new value is inserted.
+    cache.get("1");
+
+    cache.insert("4", new DisposableString("D"));
+    expect(cache.size).toBe(3);
+    expect(cache.get("1")?.value).toBe("A");
+    expect(cache.get("2")?.value).toBe(undefined);
+    expect(cache.get("3")?.value).toBe("C");
+    expect(cache.get("4")?.value).toBe("D");
+  });
+
+  it("replaces values when reinserted", () => {
+    const cache = new DataCache<DisposableString>(3);
+    cache.insert("1", new DisposableString("A"));
+    expect(cache.get("1")?.value).toBe("A");
+
+    cache.insert("1", new DisposableString("AA"));
+    expect(cache.get("1")?.value).toBe("AA");
+  });
+
+  it("accounts for entry size when evicting to maintain max size", () => {
+    const cache = new DataCache<DisposableString>(3);
+    cache.insert("1", new DisposableString("A"), 1);
+    expect(cache.size).toBe(1);
+    cache.insert("2", new DisposableString("B"), 2);
+    expect(cache.size).toBe(3);
+
+    // Insert a third value that has a size of 2. Key 1 is the oldest value,
+    // but because key 2 also has a size of 2, both must be
+    // evicted to make room for the new value.
+    cache.insert("3", new DisposableString("C"), 2);
+    expect(cache.size).toBe(2);
+    expect(cache.get("1")?.value).toBe(undefined);
+    expect(cache.get("2")?.value).toBe(undefined);
+    expect(cache.get("3")?.value).toBe("C");
+  });
+
+  it("allows entry sizes larger than capacity", () => {
+    const cache = new DataCache<DisposableString>(3);
+    cache.insert("1", new DisposableString("A"), 4);
+    expect(cache.size).toBe(4);
+    expect(cache.get("1")?.value).toBe("A");
+
+    // Inserting another large value should evict the first value.
+    cache.insert("2", new DisposableString("B"), 18);
+    expect(cache.size).toBe(18);
+    expect(cache.get("1")?.value).toBe(undefined);
+    expect(cache.get("2")?.value).toBe("B");
+  });
+
+  it("allows values that are non-disposable", () => {
+    const cache = new DataCache<string[]>(2);
+    cache.insert("1", ["A"]);
+    cache.insert("2", ["B", "B"]);
+    expect(cache.size).toBe(2);
```

Comment:
optional/nit: I think this test makes more sense if the cached sizes are actually proportional to the lengths of the arrays (or needs a separate test for "non-disposable with sizes") -- semi-related to my comment about the cache being able to ask the object for its size automatically.

Reply:
Gotcha. I think I'll just revert these to being length 1, since there's already a test for different sizes. I don't expect the disposal code to interact with size at all.